hpy/devel/include/common/hpyfunc.h

```diff
@@ -70,6 +72,19 @@ typedef enum {
     enum { SYM##_sig = SIG };              \
     _HPyFunc_TRAMPOLINE_##SIG(SYM, IMPL)
 
+typedef struct {
+    void *buf;
+    HPy obj;        /* owned reference */
```

Comment:
I think it would be better to remove `obj` from `HPy_buffer` and store it in `internal` for the CPython HPy implementations. This avoids the interface needing to be fixed when we figure out how to deal with long-lived references to objects and hides an implementation detail.

I'm not actually sure why `obj` exists on CPython -- I suspect it is implementation detail that was exposed for no good reason and is there only to allow the reference to the object to be DECREF'ed when the buffer is released.